src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs

```diff
@@ -227,12 +227,19 @@ public TypeWithAnnotations SetIsAnnotated(CSharpCompilation compilation)
                 }
             }
 
-            if (((TypeParameterSymbol)typeSymbol).TypeParameterKind == TypeParameterKind.Cref)
+            var typeParameterSymbol = (TypeParameterSymbol)typeSymbol;
+            if (typeParameterSymbol.TypeParameterKind == TypeParameterKind.Cref)
             {
                 // We always bind annotated type parameters in cref as `Nullable<T>`
                 return makeNullableT();
             }
 
+            if (typeParameterSymbol.DeclaringMethod is {} declaringMethod && (declaringMethod.IsOverride || declaringMethod.IsExplicitInterfaceImplementation))
```

Comment:
>if (typeParameterSymbol.DeclaringMethod is {} declaringMethod && (declaringMethod.IsOverride || declaringMethod.IsExplicitInterfaceImplementation)) [](start = 12, length = 147)

This condition should be applicable only to type references in a signature of a method declaration, but it looks like this isn't the case.

Reply:
For example, what would be the type of the local in this scenario (please add a test)?
```
    interface I5
    {
        void M5<T>(T value) where T : class;
    }

    class C51 : I5
    {
        void I5.M5<T>(T value)
        {
            T? x = value;
        }
    }
```

I think the more robust approach would be to force the type from the Binder, if the Binder knows that it binds types from the signature. Or, even better, to force the type in method symbol implementation. That would allow us to take a look at the presence of the constraints once we start working on step #2. 

---
In reply to: [273635920](https://github.com/dotnet/roslyn/pull/34584#discussion_r273635920) [](ancestors = 273635920)

Reply:
The binder doesn't know it binds types from the signature.

As I see it we have three options:

1. When creating the TypeParameterSymbols in SourceOrdianaryMethodSymbol, recursively check the ParameterSyntaxes to see if any TypeParameters are used as NullableTypes.

2. When binding a Type, pass in enough information for the binder to know it is binding types from the signature.

3. When binding a parameter, deep copy the type of the parameter, replacing all lazy nullable types with Nullable Types.

I think 3. is the worst option, as it is both most expensive, and most error prone. I am not sure about 1 or 2, but will go with 1 for now unless you tell me otherwise.

Reply:
I would try to do this inside method symbol. Once the return types and parameter types are bound, we can visit them and force lazy nullable types created over method type parameters to resolve towards ```Nullable<T>```. We do not need to replace the ```TypeWithAnnotations``` instances. Of course, that would require adding some new APIs to ```TypeWithAnnotations```. We would need to be able to get to the underlying type (```__defaultType```?) without triggering resolution. And we would need an API that would allow us to resolve the type towards ```Nullable<T>```.

---
In reply to: [273838023](https://github.com/dotnet/roslyn/pull/34584#discussion_r273838023) [](ancestors = 273838023)

Reply:
Given a case like this:
```csharp
    interface I
    {
        void M<T>(T? value) where T : class;
    }

    class C : I
    {
        void I.M<T>(T? value)
        {
            T? x = value;
        }
    }
```
If we implement your suggestion, we will warn on `T? x = value;` that T is an unconstrained type.
Is  this a problem?

Reply:
>If we implement your suggestion, we will warn on T? x = value; that T is an unconstrained type.
Is  this a problem?

I don't think this is a problem, assuming that we are going to get another error saying the implemented method is not found. I think this is the current behavior for the scenario when I.M is not present (i.e. the interface is empty).

---
In reply to: [274280058](https://github.com/dotnet/roslyn/pull/34584#discussion_r274280058) [](ancestors = 274280058)